### PR TITLE
Refine device context and fix GetPlace()

### PR DESCRIPTION
**Pull request by gangliao**

### CODE REVIEW

File: paddle/platform/device_context.cc [hunk 1]
```diff
@@ -20,12 +20,96 @@ Eigen::DefaultDevice* DeviceContext::get_eigen_device<Eigen::DefaultDevice>()
   return reinterpret_cast<const CPUDeviceContext*>(this)->eigen_device();
 }
 
+CPUDeviceContext::CPUDeviceContext() {
```
Comment (gangliao) [inline paddle/platform/device_context.cc]: Move implementation into `*.cc` to make the definition  clearer.

File: paddle/platform/device_context.cc [hunk 2]
```diff
@@ -20,12 +20,96 @@ Eigen::DefaultDevice* DeviceContext::get_eigen_device<Eigen::DefaultDevice>()
...
+
+CUDADeviceContext::CUDADeviceContext(GPUPlace place) : place_(place) {
+  SetDeviceId(place_.device);
+  PADDLE_ENFORCE(cudaStreamCreate(&stream_));
+  eigen_stream_.reset(new Eigen::CudaStreamDevice(&stream_));
+  eigen_device_.reset(new Eigen::GpuDevice(eigen_stream_.get()));
+}
+
+CUDADeviceContext::~CUDADeviceContext() {
+  SetDeviceId(place_.device);
```
Comment (gangliao) [inline paddle/platform/device_context.cc]: add `SetDeviceId(place_.device);` in here

File: paddle/platform/device_context.cc [hunk 3]
```diff
@@ -20,12 +20,96 @@ Eigen::DefaultDevice* DeviceContext::get_eigen_device<Eigen::DefaultDevice>()
...
+
+  if (curand_generator_) {
+    PADDLE_ENFORCE(dynload::curandDestroyGenerator(curand_generator_));
+  }
+  eigen_stream_.reset();
+  eigen_device_.reset();
+  PADDLE_ENFORCE(cudaStreamDestroy(stream_));
+}
+
+Place CUDADeviceContext::GetPlace() const { return place_; }
```
Comment (gangliao) [inline paddle/platform/device_context.cc]: Fix bug in here, return `place_`

File: paddle/platform/device_context.cc [hunk 4]
```diff
@@ -20,12 +20,96 @@ Eigen::DefaultDevice* DeviceContext::get_eigen_device<Eigen::DefaultDevice>()
   return reinterpret_cast<const CPUDeviceContext*>(this)->eigen_device();
 }
 
+CPUDeviceContext::CPUDeviceContext() {
+  eigen_device_.reset(new Eigen::DefaultDevice());
+}
+
+CPUDeviceContext::CPUDeviceContext(CPUPlace place) {
```
Comment (gangliao) [inline paddle/platform/device_context.cc]: Keep unified cpu and gpu interface

File: paddle/platform/device_context.h [hunk 5]
```diff
@@ -39,134 +39,65 @@ class DeviceContext {
...
+  GPUPlace place_;
 
-  std::unique_ptr<Eigen::CudaStreamDevice> eigen_stream_;
+ private:
   std::unique_ptr<Eigen::GpuDevice> eigen_device_;
+  std::unique_ptr<Eigen::CudaStreamDevice> eigen_stream_;
 
-  cublasHandle_t blas_handle_{nullptr};
+ private:
+  uint64_t seed_;
```
Comment (gangliao) [inline paddle/platform/device_context.h]: It's better to set seed to `uint64_t` and I will also add rand method for CPU

File: paddle/memory/memcpy.cc [hunk 6]
```diff
@@ -56,7 +56,7 @@ void Copy<platform::GPUPlace, platform::GPUPlace>(platform::GPUPlace dst_place,
                                                   const void* src, size_t num,
                                                   cudaStream_t stream) {
   if (dst_place == src_place) {
-    platform::GPUPlaceGuard g(src_place.device);
+    platform::SetDeviceId(src_place.device);
```
Comment (QiJune) [inline paddle/memory/memcpy.cc]: 这个SetDeviceId是不是要写在if外面

Comment (gangliao) [inline paddle/memory/memcpy.cc]: 这个不需要了  因为GpuMemcpyPeer函数里面的两个参数 就是相应的device id

**Pull request merged**